src/coreclr/vm/peassembly.cpp

```diff
@@ -320,12 +324,40 @@ void PEAssembly::ConvertMDInternalToReadWrite()
     }
     CONTRACTL_END;
 
-    IMDInternalImport *pOld;            // Old (current RO) value of internal import.
-    IMDInternalImport *pNew = NULL;     // New (RW) value of internal import.
-
-    // Take a local copy of *ppImport.  This may be a pointer to an RO
+    // Take a local copy of the assembly's MD import.  This may be a pointer to an RO
     //  or to an RW MDInternalXX.
-    pOld = m_pMDImport;
+    IMDInternalImport *pNew = NULL;                                 // New (RW) value of internal import.
+    IMDInternalImport *pOld = m_pMDImport;                          // Old (current RO) value of internal import.
+    IMDInternalImport *pConvertedImport = m_pConvertedMDImport;     // Pointer to a RW internal metadata that might not be actively in use.
+
+
+    if (pConvertedImport != NULL)
+    {
+        if (openForWriting && pOld != pConvertedImport)
+        {
+            // We had converted the metadata before, now we are requesting
+            // to replace the RO view we had with the converted one.
+            if (InterlockedCompareExchangeT(&m_pMDImport, pConvertedImport, pOld) == pOld)
+            {
+                //if the debugger queries, it will now see that we have RW metadata
+                m_MDImportIsRW_Debugger_Use_Only = TRUE;
+
+                // Swapped -- get the metadata to hang onto the old Internal import.
+                // Additionally, now both m_pMDImport and m_pConvertedMDImport point to
+                // the same RW MDImport. Up the references such that both have to be freed
+                // before freeing the RW copy and the RO that hangs off it.
+                pConvertedImport->AddRef();
```

Comment:
I still don't understand this part. I think we only get in here if `m_pMDImport` and `m_pConvertedMDImport` are non-null _and_ aren't the same instance. If this is true, then both `pConvertedImport` and `pOld` were/are fields on this class. I would assume then that both have correct references counts for being on this class. I do see the `pOld` being passed into `SetUserContextData()` and this could be a lifetime transfer, but if that were the case, I would expect `pOld` to need an `AddRef()` and not `pConvertedImport`.